loadbalancer: Add LBMap pressure metrics

joamaki opened the pull request:
This adds the BPF pressure gauges for all the BPF maps used in pkg/loadbalance/maps. The pressure metric is updated every 5 minutes through `NextKeyBytes` count. This is meant to be a temporary measure until we have a BPF iterator way of grabbing BPF map counts (https://github.com/cilium/cilium/pull/37927 explored this, follow up later). Since we already switched to the new load-balancing control-plane I want to make sure we have this implemented for v1.18 even if in a sub-optimal way.

The new load-balancing maps do not use `pkg/bpf` nor do they use the `WithCache` mechanism (no point since the map operation retries are handled by the reconciler (`pkg/loadbalancer/reconciler`). This means we can't use the `(*bpf.Map).WithPressureMetric` and instead need to do it this way.

The first two commits improve the BPFMapPressureGauge infra by removing the functions and moving this to be a method under `metrics.Registry`. This way we avoid the annoying `metrics.registry` promise that a lot of tests hit a timeout with, but of course since lot of the maps are still globals I had to snake through the `*metrics.Registry` to these. The `WithPressureGauge` will be a no-op with a `nil` registry so we can use that in tests or cilium-dbg.

joamaki committed:
loadbalancer: Add BPF map pressure metrics

Periodically (every 5 minutes) update the BPF map pressure metrics for all opened LB maps.
This is on by default but can be disabled with '--lb-pressure-metrics-interval=0' (in
helm can be set via 'extraConfig').

This is somewhat different to the "WithCache()" way of doing pressure metrics
as we don't have an easy way of counting how many entries should be in the BPF
map (doable, but then we'd need to match the logic in bpf_reconciler.go exactly).

As a more foolproof way we instead get the number of keys in each BPF map
and report the ratio of that to MaxEntries. This of course won't go beyond
1.0.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
metrics: Add Registry.NewBPFPressureMetricsGauge method

This allows registering pressure metrics without going through the registry
global variable.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

mhofstetter reviewed (approved):
`envoy` :heavy_check_mark:

joestringer commented on pkg/loadbalancer/experimental/pressure_metrics.go:
```diff
@@ -0,0 +1,92 @@
+	updateInterval := 5 * time.Minute
+	if pm.TestConfig != nil {
+		updateInterval = 10 * time.Millisecond
+	}
```
nit: References to this `TestConfig` structure seem a little odd to me in production code. This seems like a parameter that could be part of the config. I guess if we took this to the logical conclusion we could even express the top level config flag as an update interval where `0` means off, default is `5*time.Minute`, then the test framework just passes `10*time.Millisecond` in.

joamaki replied:
Ah yes, much better. I can even just remove the enable pressure metrics and just replace that with e.g. `lb-pressure-metrics-interval` flag.

joestringer commented on pkg/ebpf/map.go:
```diff
@@ -216,3 +216,29 @@ func (m *Map) IsEmpty() bool {
+			key, err = m.NextKeyBytes(key)
+		}
+		if key == nil || err != nil {
+			break
```
Should we record the count if an error occurs? At a glance, `NextKeyBytes()` seems like it returns nil key and value when it reaches end of the map, so it should be possible to log and avoid recording an incorrect count on failure? Maybe set the count to 0 so it's obviously incorrect as a subtle sign that something went really wrong with the map interactions?

joamaki replied:
Good catch. I think changing signature to `Count() (int, error)` actually makes more sense and then returning `(0, err)` when error is caught.

joestringer reviewed (approved):
Couple of minor notes but otherwise LGTM.

giorio94 commented on pkg/ciliumenvoyconfig/exp_test.go:
Nit. (referring to the commit message, not this file)

> This is on by default but can be disabled with '--enable-lb-pressure-metrics=false' (in
helm can be set via 'extraConfig').

It looks like the flag is actually `lb-pressure-metrics-interval`.

joamaki replied:
Agh. Always forget to update the commit messages after code changes.

giorio94 commented on pkg/loadbalancer/experimental/pressure_metrics.go:
```diff
@@ -0,0 +1,94 @@
+}
+
+func (pm *pressureMetrics) report(ctx context.Context) error {
+	var errs error
```
`errs` is only accumulated, but never returned.

giorio94 reviewed (approved):
Thanks. A couple of nits inline.

derailed reviewed (approved):
@joamaki Nice work!

joamaki committed:
treewide: Call NewBPFMapPressureGauge via Registry

This allows removing [metrics.NewBPFMapPressureGauge] and reduces
the uses we have of the [metrics.registry] promise. This will also
speed up tests as we won't have tests timing out trying to grab the
registry via the promise on map pressure updates.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

christarazi commented on pkg/loadbalancer/maps/pressure_metrics.go:
```diff
@@ -0,0 +1,95 @@
+		if ctx.Err() != nil {
+			break
+		}
+		if count, err := m.Count(); err != nil {
```
The CT map BPF pressure metric uses batch ops to count the elements in the map. Curious why the same approach is not taken here

joamaki replied:
Ah wasn't aware of this! I'm not using `pkg/bpf` so no access to `NewBatchIterator` or `CountAll` though. Could look at whether we can make those available when using plain `ebpf.Map`. Or I could also consider switching `pkg/loadbalancer/maps` over to `pkg/bpf`.

joamaki replied:
I'm going to switch to `pkg/bpf` (in another PR) as `pkg/ebpf` is being removed and will consider using the batch ops for this.

joamaki replied:
https://github.com/cilium/cilium/pull/39210

Will wiat for that to merge and will then update this PR.

joamaki replied:
Hmm, BatchIterator requires concrete types but I don't have them in `pressure_metrics.go` so I can't instantiate it. I'll look at maybe doing a `(*bpf.).Count` method that uses batch lookups underneath instead.

joamaki committed:
bpf: Add BatchCount method

BatchCount() uses BatchLookup to count up the number of elements in the BPF map.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

christarazi reviewed (approved):
New `BatchCount()` approach LGTM. We can always optimize it in the future

tommyp1ckles commented on pkg/bpf/map_linux.go:
```diff
@@ -132,6 +132,63 @@ func (m *Map) Type() ebpf.MapType {
+				} else {
+					chunkSize *= 2
+				}
+				keys = make(nopDecoder, chunkSize)
```
hmmm started playing around with this, wondering if we could retool the general IterateAll function to accommodate this use case to avoid some of the code duplication.

joamaki replied:
Yep we probably could. OK to do as a follow-up?

tommyp1ckles replied:
Sounds good!

tommyp1ckles commented on pkg/loadbalancer/maps/pressure_metrics.go:
```diff
@@ -0,0 +1,105 @@
+		m := *desc.target
+		name := m.Name()
+
+		if m.Type() != ebpf.Hash {
```
Do we need pressure on any LPMTrie maps?

joamaki replied:
Ah yeah I guess we could report the source range map pressure too. Good catch.

joamaki replied:
```
time=2025-05-09T15:13:09Z level=error source=/go/src/github.com/cilium/cilium/vendor/github.com/cilium/hive/job/timer.go:203 msg="Timer job errored" module=agent.controlplane.loadbalancer-maps name=pressure-metrics-reporter func=maps.(*pressureMetrics).report error="count on cilium_lb4_source_range failed: failed to iterate map: map batch lookup: not supported\ncount on cilium_lb6_source_range failed: failed to iterate map: map batch lookup: not supported" (1 occurrences)
```

Ah this won't work for LPMTries. I'll revert and only check hash maps.

tommyp1ckles commented on pkg/loadbalancer/maps/pressure_metrics.go:
```diff
@@ -0,0 +1,105 @@
+			errs = errors.Join(errs, fmt.Errorf("count on %s failed: %w", name, err))
+		} else {
+			pm.getGauge(name).Set(
+				float64(count) / float64(m.MaxEntries()),
```
nit: could add an extra guard here against division by zero in which case we'd set gauge to be +inf (not sure what the implications of this would be).

joamaki replied:
good point, added.

tommyp1ckles commented on pkg/loadbalancer/maps/pressure_metrics.go:
```diff
@@ -0,0 +1,105 @@
+		if ctx.Err() != nil {
+			break
+		}
+		if count, err := m.BatchCount(); err != nil {
```
Hmm, just a thought -  I wonder if in some cases, such as in scenarios with larger maps configured, we might see this loop start to take a while (i.e. exceed the prometheus scrape time).

joamaki replied:
Even the `NextKeyBytes()` style iteration that does one syscall per key can do millions of entries per second. And note that we are just updating the metric variables (essentially updating an atomic float64) and not implementing a metrics `Observer` so this cannot block anything.
